magma/backend/mlir/xmr_utils.py

```diff
@@ -0,0 +1,84 @@
+from magma.backend.mlir.magma_common import (
+    value_or_value_wrapper_to_tree as magma_value_or_value_wrapper_to_tree
+)
+from magma.ref import TupleRef, ArrayRef
+
+
+def _get_path(tree, node):
+    predecessors = tree.predecessors(node)
+    try:
+        parent = next(predecessors)
+    except StopIteration:
+        return [node]
+    try:
+        next(predecessors)
+        raise RuntimeError()
+    except StopIteration:
+        pass
+    return _get_path(tree, parent) + [node]
+
+
+def _path_to_string(path, seperator):
+    path = [str(path[0])] + [str(v.name.index) for v in path[1:]]
+    return seperator.join(path)
+
+
+def _get_leaf_descendants(tree, node, include_self=False):
+    is_leaf = True
+    for succ in tree.successors(node):
+        is_leaf = False
+        yield from _get_leaf_descendants(tree, succ, include_self=True)
+    if is_leaf and include_self:
+        yield node
```

Comment:
Could just use succ as the flag
```suggestion
    succ = None
    for succ in tree.successors(node):
        yield from _get_leaf_descendants(tree, succ, include_self=True)
    if succ is not None and include_self:
        yield node
```